Avoid paint on popover when hovering content

corentin-gautier opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
Popovers use outline instead of border but it seems to trigger paint when hovering elements inside the popover

## Why?
Using border avoids triggering paint which is better for performances

## How?
Replace the use of outline by border

## Testing Instructions
1. Open the developer console and enable paint flashing
2. Open a color palette popover
3. Hover over a circular picker
4. There should not be paint of the entire popover area

## Screenshots

### Before
https://user-images.githubusercontent.com/4660731/204803471-98ed7757-516c-4960-8ada-e6b394b9621a.mov

### After
https://user-images.githubusercontent.com/4660731/204803498-fba2f762-97b9-4a5a-930f-1150dbefae1e.mov

ciampo commented:
@jasmussen do you have any context to why we may want to use `outline` instead of `border` for `Popover`? This PR is proposing a rendering perf improvement, but it uses `border` instead of `outline`.

jasmussen commented:
Based on this comment here, https://github.com/WordPress/gutenberg/pull/40740/files#diff-7895f8092f907d624abdbe647e2201f1cebe9198cd3fae6ba759b70a4c0bfef4R16, it appears to be related to popover positioning computations. @youknowriad might have more context, but it sounds like the math that positions the inserter benefits from the outline not counting towards the block footprint.

An alternate solution, not sure if it affects perf, is to use an inner box-shadow for the border. That should theoretically accomplish the same.

ciampo reviewed (commented):
As @jasmussen says, we can't use `border` because it affects the computed position for the `Popover`

| This PR | `trunk` |
|---|---|
| ![Screenshot 2022-12-02 at 10 32 40](https://user-images.githubusercontent.com/1083581/205266598-cb00ccd6-9159-4750-9156-064f8394d048.png) | ![Screenshot 2022-12-02 at 10 32 45](https://user-images.githubusercontent.com/1083581/205266645-e22ea392-2360-43dc-8f28-f203963efde9.png) |


I tried using `box-shadow` on my machine as suggested:

<details>

<summary>Click to see the code changes I applied</summary>

```diff
diff --git a/packages/components/src/popover/style.scss b/packages/components/src/popover/style.scss
index f91f08b84f..c99034234d 100644
--- a/packages/components/src/popover/style.scss
+++ b/packages/components/src/popover/style.scss
@@ -1,5 +1,10 @@
 $arrow-triangle-base-size: 14px;
 
+$shadow-popover-border-default: 0 0 0 $border-width $gray-400;
+$shadow-popover-border-default-alternate: 0 0 0 $border-width $gray-900;
+$shadow-popover-border-top-only: 0 #{-$border-width} 0 0 $gray-400;
+$shadow-popover-border-top-only-alternate: 0 #{-$border-width} 0 $gray-900;
+
 .components-popover {
 	z-index: z-index(".components-popover");
 
@@ -15,16 +20,16 @@ $arrow-triangle-base-size: 14px;
 
 .components-popover__content {
 	background: $white;
-	border: $border-width solid $gray-400;
-	box-shadow: $shadow-popover;
+	// Using `box-shadow` instead of `border` to avoid impacting
+	// popover computations.
+	box-shadow: $shadow-popover-border-default, $shadow-popover;
 	border-radius: $radius-block-ui;
 	box-sizing: border-box;
 	width: min-content;
 
 	// Alternate treatment for popovers that put them at elevation zero with high contrast.
 	.is-alternate & {
-		border-color: $gray-900;
-		box-shadow: none;
+		box-shadow: $shadow-popover-border-default-alternate;
 	}
 
 	// A style that gives the popover no visible ui.
@@ -41,7 +46,11 @@ $arrow-triangle-base-size: 14px;
 		overflow-y: visible;
 		width: auto;
 		border: none;
-		border-top-color: $gray-900;
+		box-shadow: $shadow-popover-border-top-only, $shadow-popover;
+	}
+
+	.components-popover.is-expanded.is-alternate & {
+		box-shadow: $shadow-popover-border-top-only-alternate, $shadow-popover;
 	}
 }
 ```
</details>

From a quick look:

- paint flashes are still reduced as per this PR's intention
- Popover looks like on `trunk` (and the new border still doesn't affect the Popover's position)

@corentin-gautier , do you mind applying the suggested changes from the code snippet above and check if it achieves the rendering perf improvement that you had in mind?

corentin-gautier commented:
@ciampo @jasmussen box-shadow doesn't trigger paint 👍

ciampo commented:
Before merging, do you mind adding a CHANGELOG entry for the changes in the `@wordpress/components` package, as suggested by Aaron in [this other PR](https://github.com/WordPress/gutenberg/pull/46196#pullrequestreview-1200288177)?

ciampo commented:
@mirka or @tyxla do you mind giving this PR a final round of review? I'm not going to be around in the upcoming weeks

tyxla reviewed (commented):
This looks pretty good! Thanks @corentin-gautier 🙌 

Let's have a CHANGELOG entry and I'm happy to ship it afterwards.

corentin-gautier commented:
> This looks pretty good! Thanks @corentin-gautier 🙌
> 
> Let's have a CHANGELOG entry and I'm happy to ship it afterwards.

Done !

tyxla commented:
> > This looks pretty good! Thanks @corentin-gautier 🙌
> > Let's have a CHANGELOG entry and I'm happy to ship it afterwards.
> 
> Done !

Thanks, @corentin-gautier, but could you please rebase the branch to the latest `trunk` as well? It seems to have some conflicts currently.

Commit:
Avoid paint on popover when hovering content

remove border on unstyled/expanded

use box-shadow instead of border

remove shadow on alternate

update changelog

corentin-gautier commented:
@tyxla and done ;)

tyxla commented:
Thanks again @corentin-gautier 🙌 Let me ship this for you.

tyxla reviewed (approved):
✅
